### PR TITLE
SECFingerprintTransformer implementation

### PR DESCRIPTION
**Pull request by rafalbachorz**
As discussed on Slack - simple implementation exploiting the RDKit capabilities.

### CODE REVIEW

File: scikit_mol/transformers.py [hunk 1]
```diff
@@ -205,6 +206,31 @@ def _mol2fp(self, mol):
...
+                 min_radius:int=1, length:int=2048, n_permutations:int=0, seed:int=0):
+        self.radius = radius
+        self.rings = rings
+        self.isomeric = isomeric
+        self.kekulize = kekulize
+        self.min_radius = min_radius
+        self.nBits = length
+        self.n_permutations = n_permutations
+        self.seed = seed
+        self.secfp_encoder = rdMHFPFingerprint.MHFPEncoder(self.n_permutations, self.seed)
```
Comment (EBjerrum) [inline scikit_mol/transformers.py:221]: I don't think this will be scikit-learn compatible. scikit-learn expects during some routines to be able to set arguments via .set_params(). There it could set .seed and n_permutations, but that would not update the secfp_encoder.

You will probably need a @property and maybe corresponding setter

File: tests/test_fptransformers.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 from rdkit import Chem
 from fixtures import mols_list, smiles_list, fingerprint
 from sklearn import clone
-from scikit_mol.transformers import MorganTransformer, MACCSTransformer, RDKitFPTransformer, AtomPairFingerprintTransformer, TopologicalTorsionFingerprintTransformer
+from scikit_mol.transformers import MorganTransformer, MACCSTransformer, RDKitFPTransformer, AtomPairFingerprintTransformer, TopologicalTorsionFingerprintTransformer, SECFingerprintTransformer
```
Comment (EBjerrum) [inline tests/test_fptransformers.py:7]: Great, you also took the effort to add to the unit tests :-)

Comment (EBjerrum) [reply tests/test_fptransformers.py:7]: I'll test in the weekend

**Review by EBjerrum — CHANGES_REQUESTED** [review]

File: scikit_mol/transformers.py (hunk 1 above)
Comment (rafalbachorz) [reply scikit_mol/transformers.py:221]: Thanks, I will fix this and let you know.

Comment (EBjerrum) [reply scikit_mol/transformers.py:221]: @seed.setter and @n_permutations.setter could create the object. That would create it twice during init, but that's probably more efficient than thousands of times during fingerprint transforms.

Comment (rafalbachorz) [reply scikit_mol/transformers.py:221]: My intention was to avoid the situation that each time the _mol2fp method is called there is the necessity of creating an encoder/generator object. If this encoder object does not depend on the arguments of the _mol2fp (and in the the cases we have so far it doesn't) then it can be created once during the instantiation of a transformer object. That's the reason I have put that to the class constructor. But it might be that it is not fully compatible with the scikit-learn workflows.

Comment (EBjerrum) [reply scikit_mol/transformers.py:221]: I understand why you do put it in init, but it's not acceptable as we strive for full scikit-learn compatibility. Yes the pipeline works, but the hyperparameter optimizations will probably not do as intended, if you try to optimize seed or n_permutations.

Look here under set_params() https://scikit-learn.org/stable/developers/develop.html, it wont work in grid_searches and such.

You must be able to call .set_params(seed=Something_new), and then that should make a new instance of the FP generator object. Currently it wont.

Something like 

@seed.setter
def seed(self, seed):
    self.seed = seed
    self.secfp_encoder = rdMHFPFingerprint.MHFPEncoder(self.n_permutations, self.seed)

Will get around this. Problem is the MHFPEncoder depends on two properties from the object (which can probably be solved with hidden properties like self._seed and self._n_permutations with default values).

Alternatively a cached getter on self.secfp_encoder, but the cache needs to be cleared when either seed or n_permutations is needed, not sure if there's a mechanism for that on cached functions.

Or the setter of seed and n_permutations sets a property ._redo_encoder, that are then used to figure out if the encoder need to be recreated in the .getter of that property.

Maybe there's yet other solutions that are even more efficient.

Simplest codewise is to put the creation of the MHFPEncoder in the ._molfp method, but what is the computational overhead compared to the other implementations I have sketched? I don't know.

Comment (EBjerrum) [reply scikit_mol/transformers.py:221]: It has already been necessary to use getters and setters in RDKitFPTransformer and Desc2DTransformer, so for consistency that is the current way.

Comment (EBjerrum) [reply scikit_mol/transformers.py:221]: But for the RDKitFPTransformer it currently recreates the fp encoder object every time. It depends on a lot of settings, so it will be a lot of boilerplate code to make it recreate it everytime a property is set differently, but eventually it may need to be handled.

Comment (rafalbachorz) [reply scikit_mol/transformers.py:221]: Thanks for this elaborated response. I see the points, we need to have a mechanism that whenever the seed/n_permutations are changed the encoder instance needs to be refreshed. As we have here only two such parameters it is probably worth elaborating slightly the setters, as you suggest, in order to recreate the encoder. In the case of the RDKitFPTransformer it is probably worth doing the same, even though there are more setters/getters to be written, unless there is some other, less code-demanding way of doing that.
So, how shall we proceed further? Shall I make the relevant corrections and then make another PR?

Comment (EBjerrum) [reply scikit_mol/transformers.py:221]: Yes, please do. That would be great. If you update the code and make a new pull request, this one will get updated. 

I'll look at trying to make a unit-test that cover the issue when .set_params() does not do as intented.

I'll add on the TODO list that we should look at the RDKitFPTransformer, but for now, it has the correct behaviour, although it may be less computationally efficient than reusing the encoder.

**Comment by rafalbachorz** [comment]
Please let me know if you see an update of my PR.

**Comment by EBjerrum** [comment]
Yes, I see it.

**Comment by EBjerrum** [comment]
Looks perfect

**Comment by rafalbachorz** [comment]
Great - thanks!
